### DISCUSSION TITLE
Using jupyter notebook causes `RuntimeError: asyncio.run() cannot be called from a running event loop`

### DISCUSSION THREAD

**Discussion by tyehuoPHD** [Rapptz/discord.py#9271] [Q&A]
# fact and question
The code works when I call the `.py` file using `python test.py` on command line. It does not work when I executed the code inside the cell of jupyter notebook.

How do i modify my code such that it would work in jupyter notebook?

# Here's my code. 
```python
import os
import discord
import asyncio
from dotenv import load_dotenv

load_dotenv()
TOKEN = os.getenv('DISCORD_TOKEN')
print(TOKEN)

intents = discord.Intents.default()
intents.message_content = True

client = discord.Client(intents=intents,proxy='http://127.0.0.1:7890')

@client.event
async def on_ready():
    print(f'We have logged in as {client.user}')

@client.event
async def on_message(message):
    if message.author == client.user:
        return

    if message.content.startswith('$hello'):
        await message.channel.send('Hello!')

client.run(TOKEN)
```

# Here's my errors
```log
RuntimeError                              Traceback (most recent call last)
Cell In[25], line 34
     31     if message.content.startswith('$hello'):
     32         await message.channel.send('Hello!')
---> 34 client.run(<TOKEN I HAVE DELETED>)
     36 # asyncio.run()
     37 
     38 
   (...)
     91 # async def on_ready():
     92 #     print(f'We have logged in as {client.user}')

File c:\Users\tyehu\scoop\apps\pyenv\current\pyenv-win\versions\3.11.2\Lib\site-packages\discord\client.py:828, in Client.run(self, token, reconnect, log_handler, log_formatter, log_level, root_logger)
    820     utils.setup_logging(
    821         handler=log_handler,
    822         formatter=log_formatter,
    823         level=log_level,
    824         root=root_logger,
    825     )
    827 try:
--> 828     asyncio.run(runner())
    829 except KeyboardInterrupt:
    830     # nothing to do here
    831     # `asyncio.run` handles the loop cleanup
...
    187         "asyncio.run() cannot be called from a running event loop")
    189 with Runner(debug=debug) as runner:
    190     return runner.run(main)

RuntimeError: asyncio.run() cannot be called from a running event loop
```
# My investigation 

I know `RuntimeError: asyncio.run() cannot be called from a running event loop` is resulted from the use of jupyter notebook executing the code.

Referencing this webpage [asyncio-run-cannot-be-called-from-a-running-event-loop-when-using-jupyter](https://stackoverflow.com/questions/55409641/asyncio-run-cannot-be-called-from-a-running-event-loop-when-using-jupyter-no), I concluded: 

The [asyncio.run()](https://docs.python.org/3.7/library/asyncio-task.html#asyncio.run) documentation says: 
> This function [cannot](https://github.com/python/cpython/blob/3.8/Lib/asyncio/runners.py#L32-L34) be called when another asyncio event loop is running in the same thread.`

jupyter ([IPython ≥ 7.0](https://blog.jupyter.org/ipython-7-0-async-repl-a35ce050f7f7)) is already running an event loop:
>  You can now use async/await at the top level in the IPython terminal and in the notebook, it should — in most of the cases — “just work”. Update IPython to version 7+, IPykernel to version 5+, and you’re off to the races.

This function [cannot](https://github.com/python/cpython/blob/3.8/Lib/asyncio/runners.py#L32-L34) be called when another asyncio event loop is running in the same thread.

**Comment by 2br-2b** [comment]
Could `asyncio.create_task(client.run(TOKEN))` work? I believe that will run it in the current event loop.

It would either be that or [one of these solutions](https://stackoverflow.com/questions/34499400/how-to-add-a-coroutine-to-a-running-asyncio-loop):
```python
loop = asyncio.get_event_loop()
loop.run_until_complete(client.run(TOKEN))
loop.close()
```

**Reply by Gobot1234** [reply]
client.run is not a coroutine so won't work but client.start might
